File: apps/web/src/api/gradebook.ts
import client from './client';

export interface GradeCategoryData {
  id: string;
  tenant_id: string;
  subject_id: string;
  group_id: string;
  term_id: string;
  name: string;
  weight: number;
  created_at: string;
}

export interface AssessmentData {
  id: string;
  tenant_id: string;
  subject_id: string;
  group_id: string;
  term_id: string;
  category_id: string | null;
  teacher_id: string;
  title: string;
  description: string | null;
  date: string;
  max_score: number;
  coefficient: number;
  is_published: boolean;
  created_at: string;
}

export interface GradeData {
  id: string;
  assessment_id: string;
  student_id: string;
  score: number | null;
  is_absent: boolean;
  is_exempt: boolean;
  comment: string | null;
  created_at: string;
}

export interface SubjectAverage {
  subject_id: string;
  subject_name: string;
  average: number | null;
  assessment_count: number;
}

export interface StudentAveragesData {
  student_id: string;
  term_id: string | null;
  averages: SubjectAverage[];
  general_average: number | null;
}

export interface GradeInput {
  student_id: string;
  score: number | null;
  is_absent: boolean;
  is_exempt: boolean;
  comment: string | null;
}

// Grade Categories
export function getGradeCategories(params: { group_id: string; subject_id: string; term_id: string }) {
  return client.get<GradeCategoryData[]>('/v1/gradebook/categories', { params });
}

export function createGradeCategory(data: { subject_id: string; group_id: string; term_id: string; name: string; weight?: number }) {
  return client.post<GradeCategoryData>('/v1/gradebook/categories', data);
}

// Assessments
export function getAssessments(params: { group_id?: string; subject_id?: string; term_id?: string } = {}) {
  return client.get<AssessmentData[]>('/v1/gradebook/assessments', { params });
}

export function createAssessment(data: {
  subject_id: string;
  group_id: string;
  term_id: string;
  category_id?: string;
  title: string;
  description?: string;
  date: string;
  max_score?: number;
  coefficient?: number;
}) {
  return client.post<AssessmentData>('/v1/gradebook/assessments', data);
}

// Grades
export function getAssessmentGrades(assessmentId: string) {
  return client.get<GradeData[]>(`/v1/gradebook/assessments/${assessmentId}/grades`);
}

export function bulkCreateGrades(assessmentId: string, grades: GradeInput[]) {
  return client.post<GradeData[]>('/v1/gradebook/grades/bulk', { grades }, { params: { assessment_id: assessmentId } });
}

// Student Averages
export function getStudentAverages(studentId: string, params: { term_id?: string } = {}) {
  return client.get<StudentAveragesData>(`/v1/gradebook/students/${studentId}/averages`, { params });
}

File: apps/api/app/db/models/gradebook.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, Column, Date, DateTime, ForeignKey, Numeric, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.db.base import Base, TenantMixin


class GradeCategory(Base, TenantMixin):
    __tablename__ = "grade_categories"

    subject_id = Column(UUID(as_uuid=True), ForeignKey("subjects.id"), nullable=False)
    group_id = Column(UUID(as_uuid=True), ForeignKey("groups.id"), nullable=False)
    term_id = Column(UUID(as_uuid=True), ForeignKey("terms.id"), nullable=False)
    name = Column(String(255), nullable=False)
    weight = Column(Numeric(5, 2), nullable=False, default=1)

    assessments = relationship("Assessment", back_populates="category", cascade="all, delete-orphan")


class Assessment(Base, TenantMixin):
    __tablename__ = "assessments"

    subject_id = Column(UUID(as_uuid=True), ForeignKey("subjects.id"), nullable=False)
    group_id = Column(UUID(as_uuid=True), ForeignKey("groups.id"), nullable=False)
    term_id = Column(UUID(as_uuid=True), ForeignKey("terms.id"), nullable=False)
    category_id = Column(UUID(as_uuid=True), ForeignKey("grade_categories.id"), nullable=True)
    teacher_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    date = Column(Date, nullable=False)
    max_score = Column(Numeric(5, 2), nullable=False, default=20)
    coefficient = Column(Numeric(5, 2), nullable=False, default=1)
    is_published = Column(Boolean, default=False)
    source_activity_id = Column(UUID(as_uuid=True), nullable=True, index=True)

    category = relationship("GradeCategory", back_populates="assessments")
    grades = relationship("Grade", back_populates="assessment", cascade="all, delete-orphan")


class Grade(Base):
    __tablename__ = "grades"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    assessment_id = Column(UUID(as_uuid=True), ForeignKey("assessments.id"), nullable=False)
    student_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    score = Column(Numeric(5, 2), nullable=True)
    is_absent = Column(Boolean, default=False)
    is_exempt = Column(Boolean, default=False)
    comment = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    assessment = relationship("Assessment", back_populates="grades")
